File: tests/fixtures/realistic/express/routes/accounts.ts
import express from "express";
import { PrismaClient } from "@prisma/client";
import { audit, requireAdmin, requireAuth, requirePermission, requireTenant } from "../app";
import { createAccount, deleteAccount, updateAccount } from "../services/accounts";

const prisma = new PrismaClient();
const router = express.Router();

function dynamicPolicyCheck(name: string) {
  return name === "accounts.update";
}

router.get("/:accountId", requireAuth, (req: express.Request, res: express.Response) => {
  res.json({ id: req.params.accountId });
});

router.post("/", requireAuth, audit, async (req: express.Request, res: express.Response) => {
  await prisma.account.create({
    data: { ownerId: req.user.id, name: req.body.name },
  });
  res.json({ created: true });
});

router.patch(
  "/:accountId",
  requirePermission("accounts.write"),
  async (req: express.Request, res: express.Response) => {
    if (!dynamicPolicyCheck("accounts.update")) {
      return res.sendStatus(403);
    }
    await updateAccount(req.params.accountId, req.body);
    res.json({ updated: true });
  },
);

router.delete("/:accountId", requireAdmin, async (req: express.Request, res: express.Response) => {
  await deleteAccount(req.params.accountId);
  res.sendStatus(204);
});

router.post("/service", requireAuth, async (req: express.Request, res: express.Response) => {
  await createAccount(req.user.id, req.body.name);
  res.json({ created: true });
});

router.post("/dynamic-service", requireAuth, (req: express.Request, res: express.Response) => {
  return accountServiceClient.create(req.user.id, req.body.name);
});

router.get("/tenant/:tenantId", requireTenant, (req: express.Request, res: express.Response) => {
  res.json({ tenantId: req.params.tenantId });
});

export default router;

File: tests/fixtures/realistic/fastapi/app/routers/accounts.py
from fastapi import APIRouter, Depends
from sqlalchemy.orm import Session

from app.models import Account
from app.services.accounts import create_account, delete_account, update_account
from main import can_edit_account, dynamic_policy_check, require_admin, require_user

router = APIRouter(prefix="/accounts")


@router.get("/{account_id}")
def read_account(account_id: str, user=Depends(require_user)):
    return {"id": account_id, "user": user["id"]}


@router.post("")
def create_account_route(session: Session, user=Depends(require_user)):
    account = Account(name="direct", owner_id=user["id"])
    session.add(account)
    session.commit()
    return {"created": True}


@router.patch("/{account_id}")
def update_account_route(account_id: str, user=Depends(can_edit_account)):
    if not dynamic_policy_check("accounts.update"):
        raise PermissionError("policy denied")
    return update_account(account_id, {"name": "renamed"})


@router.delete("/{account_id}")
def delete_account_route(account_id: str, user=Depends(require_admin)):
    return delete_account(account_id)


@router.post("/service")
def service_create_account(user=Depends(require_user)):
    return create_account(user["id"])


@router.post("/dynamic-service")
def dynamic_service_create(user=Depends(require_user)):
    return account_services["create"](user["id"])
